Utils/Nginx/config.h

```diff
@@ -0,0 +1,207 @@
+    directives.push_back(std::forward<T>(directive));
+    return Value<T>(directives.back());
+  }
+};
+
+// clang-format off
+CURRENT_STRUCT(LocationDirective, BlockDirective) {
+  CURRENT_DEFAULT_CONSTRUCTOR(LocationDirective) : SUPER("location") {}
+  CURRENT_CONSTRUCTOR(LocationDirective)(const std::string& location) : SUPER("location", location) {}
+
+  LocationDirective& AddNestedLocation(const std::string& location) {
+    return SUPER::AddDirective(LocationDirective(location));
+  }
+};
+// clang-format on
+
+inline LocationDirective DefaultLocationDirective(const std::string& location) {
+  LocationDirective result(location);
+  result.AddSimpleDirective("proxy_set_header", "X-Real-IP $remote_addr");
+  result.AddSimpleDirective("proxy_set_header", "X-Forwarded-For $proxy_add_x_forwarded_for");
+  result.AddSimpleDirective("proxy_pass_request_headers", "on");
+  return result;
+}
+
+// clang-format off
+CURRENT_STRUCT(ServerDirective, BlockDirective) {
+  CURRENT_FIELD(port, uint16_t, 0u);
+  CURRENT_DEFAULT_CONSTRUCTOR(ServerDirective) : SUPER("server") {}
+  CURRENT_CONSTRUCTOR(ServerDirective)(uint16_t port) : SUPER("server"), port(port) {
+    SUPER::AddSimpleDirective("listen", current::ToString(port));
+  }
+
+  LocationDirective& AddLocation(const std::string& location) {
+    return SUPER::AddDirective(LocationDirective(location));
+  }
+
+  LocationDirective& AddDefaultLocation(const std::string& location) {
+    return SUPER::AddDirective(DefaultLocationDirective(location));
+  }
+};
+
+CURRENT_STRUCT(HttpDirective, BlockDirective) {
+  CURRENT_DEFAULT_CONSTRUCTOR(HttpDirective) : SUPER("http") {}
+
+  ServerDirective& AddServer(uint16_t port) {
+    for (const auto& d : SUPER::directives) {
+      if (Exists<ServerDirective>(d) && Value<ServerDirective>(d).port == port) {
+         CURRENT_THROW(PortAlreadyUsedException(port));
+      }
+    }
+    return SUPER::AddDirective(ServerDirective(port));
+  }
+};
+// clang-format on
+
+using full_config_directives_t = Variant<SimpleDirective, BlockDirective>;
+
+CURRENT_STRUCT(FullConfig) {
+  CURRENT_FIELD(directives, std::vector<full_config_directives_t>);
+  CURRENT_FIELD(http, HttpDirective);
+
+  SimpleDirective& AddSimpleDirective(const std::string& name, const std::string& parameters) {
+    directives.push_back(SimpleDirective(name, parameters));
+    return Value<SimpleDirective>(directives.back());
+  }
+
+  BlockDirective& AddBlockDirective(const std::string& name, const std::string& parameters) {
+    directives.push_back(BlockDirective(name, parameters));
+    return Value<BlockDirective>(directives.back());
+  }
+};
+
+struct ConfigPrinter {
+  struct DirectivesPrinter {
+    DirectivesPrinter(std::ostream& os, size_t indent) : os_(os), indent_(indent) {}
+
+    void operator()(const SimpleDirective& simple) const {
+      PrintIndent();
+      os_ << simple.name;
+      if (!simple.parameters.empty()) {
+        os_ << ' ' << simple.parameters;
```

Comment:
What if `simple_parameters` contains spaces or quotes? Perhaps `os_ << " \"" << Escaped(simple.parameters) << '\";` is the right way? A link to the docu would be nice if it's not.